libs/base/System/Directory.idr

```diff
@@ -97,9 +98,39 @@ removeDir : HasIO io => String -> io ()
 removeDir dirName = primIO (prim__removeDir dirName)
 
 export
-dirEntry : HasIO io => Directory -> io (Either FileError String)
-dirEntry (MkDir d)
+nextDirEntry : HasIO io => Directory -> io (Either FileError (Maybe String))
+nextDirEntry (MkDir d)
     = do res <- primIO (prim__dirEntry d)
          if prim__nullPtr res /= 0
-            then returnError
-            else ok (prim__getString res)
+            then if !(getErrno) /= 0
+                    then returnError
+                    else pure $ Right Nothing
+            else do let n = prim__getString res
+                    if n == "." || n == ".."
+                       then assert_total $ nextDirEntry (MkDir d)
+                       else pure $ Right (Just n)
+
+-- This function is deprecated; to be removed after the next version bump
+export
+dirEntry : HasIO io => Directory -> io (Either FileError String)
+dirEntry d = do r <- nextDirEntry d
+                pure $ case r of
+                         Left e         => Left e
+                         Right (Just n) => Right n
+                         Right Nothing  => Left FileNotFound
+
+collectDir : HasIO io => Directory -> io (Either FileError (List String))
+collectDir d = do Right (Just n) <- nextDirEntry d
+                    | Right Nothing => pure $ Right []
+                    | Left e => pure $ Left e
+                  Right ns <- assert_total $ collectDir d
+                    | Left e => pure $ Left e
+                  pure $ Right (n :: ns)
```

Comment:
I think @Russoul has a trick to use `Compose` here so that the monad
used by the `do` block is `io . Either FileError` instead of `io` and
this way you don't need to systematically pack/unpack the `Left/Right`.

Reply:
Interesting, I didn't know Idris could do that.

I think I got how it can be done explicitly with:

```
(... >>= ...) @{Monad.Compose}
```

but result is not very readable. But I have not found to use `do` notation with that `Compose`. I. e. how to tell `do` notation to use `Monad.Compose` version of `>>=` instead of default `>>=`?

Reply:
Unfortunately, this is the best solution I can think of:
```
namespace IO.Either
  public export
  (>>=) : (IO . Either e) a -> (a -> (IO . Either e) b) -> (IO . Either e) b
  (>>=) = (>>=) @{Monad.Compose {m = IO} {t = Either e}}

  public export
  pure : a -> (IO . Either e) a
  pure = pure @{Applicative.Compose {f = IO} {g = Either e}}

collectDir d = IO.Either.do
  ...

```
It should relieve us from the packing/unpacking, but it has two noticeable flaws:
1) One has to define that namespace before being able to make use of it
2) The publicly exported `(>>=)` and `pure` can potentially result in puzzling errors in any module that depends on `System.Directory` and has ambiguous occurrences of `pure` or `(>>=)`.

The second point is quite concerning. 
So I'd refrain from applying the solution yet.

For some reason I wasn't able to make it work with a locally let-bound specialised `(>>=)`.

Reply:
Oh, it seems that this works though:
```
collectDir d = do
  let (>>=) : (IO . Either e) a -> (a -> (IO . Either e) b) -> (IO . Either e) b
      (>>=) = Prelude.(>>=) @{Monad.Compose {m = IO} {t = Either e}}
  ...   
```

Reply:
It works! Thanks! This is my version:

```
collectDir : HasIO io => Directory -> io (Either FileError (List String))
collectDir d = liftIO $ do let (>>=) : (IO . Either e) a -> (a -> (IO . Either e) b) -> (IO . Either e) b
                               (>>=) = Prelude.(>>=) @{Monad.Compose {m = IO} {t = Either e}}
                           Just n <- nextDirEntry d
                             | Nothing => pure $ Right []
                           ns <- assert_total $ collectDir d
                           pure $ Right (n :: ns)
```

I'll update the PR later.

Reply:
Updated the PR.